File: app/src/main/java/de/westnordost/osmagent/quests/ElementGeometry.java
package de.westnordost.osmagent.quests;

import java.io.Serializable;
import java.util.Collections;
import java.util.List;

import de.westnordost.osmapi.map.data.LatLon;

/** Information on the geometry of a quest */
public class ElementGeometry implements Serializable
{
	private List<List<LatLon>> outer;
	private List<List<LatLon>> inner;

	public ElementGeometry(LatLon position)
	{
		outer = Collections.singletonList(Collections.singletonList(position));
		inner = null;
	}

	public ElementGeometry(List<LatLon> positions)
	{
		outer = Collections.singletonList(positions);
		inner = null;
	}

	public ElementGeometry(List<List<LatLon>> outer, List<List<LatLon>> inner)
	{
		this.outer = outer;
		this.inner = inner;
	}


	// TODO add tests in CreateQuestMapDataHandlerTest
}

File: app/src/main/java/de/westnordost/streetcomplete/data/osm/geometry/ElementGeometry.kt
package de.westnordost.streetcomplete.data.osm.geometry

import de.westnordost.streetcomplete.data.osm.mapdata.BoundingBox
import de.westnordost.streetcomplete.data.osm.mapdata.LatLon
import de.westnordost.streetcomplete.util.math.enclosingBoundingBox
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Information on the geometry of a quest  */
@Serializable
sealed class ElementGeometry {
    abstract val center: LatLon
    // the bbox should not be serialized, his is why the bounds cannot be a (computed) property directly
    abstract fun getBounds(): BoundingBox
}

@Serializable
@SerialName("polylines")
data class ElementPolylinesGeometry(val polylines: List<List<LatLon>>, override val center: LatLon) : ElementGeometry() {
    private val bbox by lazy { polylines.flatten().enclosingBoundingBox() }
    override fun getBounds(): BoundingBox = bbox
}

@Serializable
@SerialName("polygons")
data class ElementPolygonsGeometry(val polygons: List<List<LatLon>>, override val center: LatLon) : ElementGeometry() {
    private val bbox by lazy { polygons.flatten().enclosingBoundingBox() }
    override fun getBounds(): BoundingBox = bbox
}

@Serializable
@SerialName("point")
data class ElementPointGeometry(override val center: LatLon) : ElementGeometry() {
    private val bbox by lazy { BoundingBox(center, center) }
    override fun getBounds(): BoundingBox = bbox
}
